src/graphql/data/TokenPrice.ts

```diff
@@ -0,0 +1,183 @@
+    case TimePeriod.YEAR:
+      return 'YEAR'
+    case TimePeriod.ALL:
+      return 'MAX'
+  }
+}
+
+const allQuery = graphql`
+  query TokenPriceAllQuery($contract: ContractInput!) {
+    tokenProjects(contracts: [$contract]) {
+      name
+      markets(currencies: [USD]) {
+        priceHistory1H: priceHistory(duration: HOUR) {
+          timestamp
+          value
+        }
+        priceHistory1D: priceHistory(duration: DAY) {
+          timestamp
+          value
+        }
+        priceHistory1W: priceHistory(duration: WEEK) {
+          timestamp
+          value
+        }
+        priceHistory1M: priceHistory(duration: MONTH) {
+          timestamp
+          value
+        }
+        priceHistory1Y: priceHistory(duration: YEAR) {
+          timestamp
+          value
+        }
+        priceHistoryMAX: priceHistory(duration: MAX) {
+          timestamp
+          value
+        }
+      }
+      tokens {
+        chain
+        address
+        symbol
+        decimals
+      }
+    }
+  }
+`
+
+const query = graphql`
+  query TokenPriceSingleQuery($contract: ContractInput!, $duration: HistoryDuration!) {
+    tokenProjects(contracts: [$contract]) {
+      name
+      markets(currencies: [USD]) {
+        priceHistory: priceHistory(duration: $duration) {
+          timestamp
+          value
+        }
+      }
+      tokens {
+        chain
+        address
+        symbol
+        decimals
+      }
+    }
+  }
+`
+
+export type PricePoint = { value: number; timestamp: number }
+
+function formatPrices(priceHistory: PriceHistory | undefined) {
+  return priceHistory?.filter((p): p is PricePoint => Boolean(p && p.value))
+}
+
+function extractPrices(data: TokenPriceSingleQuery['response']) {
+  return formatPrices(data?.tokenProjects?.[0]?.markets?.[0]?.priceHistory)
+}
+
+export function useTokenPriceQuery(address: string, chain: Chain, timePeriod: TimePeriod) {
+  const [prices, setPrices] = useState<PricePoint[] | undefined>(TokenAPICache.checkPriceHistory(address, timePeriod))
+  const [error, setError] = useState<any>()
+  const [isLoading, setIsLoading] = useState(!prices)
+
+  /* To be called every time the data is successfully queried */
+  const updatePrices = useCallback(
+    (data: TokenPriceSingleQuery['response']) => {
+      const priceData = extractPrices(data)
+      if (priceData) {
+        TokenAPICache.setPriceHistory(priceData, address, timePeriod)
+        setPrices(priceData)
+      }
+    },
+    [timePeriod, address]
+  )
+
+  const fetchData = useCallback(() => {
+    fetchQuery<TokenPriceSingleQuery>(environment, query, {
+      contract: {
+        address,
+        chain,
+      },
+      duration: toHistoryDuration(timePeriod),
+    }).subscribe({
+      start: () => setIsLoading(true),
+      next: updatePrices,
+      error: setError,
+      complete: () => setIsLoading(false),
+    })
+  }, [setIsLoading, address, chain, timePeriod, updatePrices, setError])
+
+  useEffect(() => {
```

Comment:
so in this case, the thing you want to be careful of is triggering multiple fetches when you are already mid-fetch. which i don't think this actually helps with. your cache should be cheap to access, and then if you don't have a hit you want to fetch the data, and while you're fetching not trigger any other fetches.

this use effect doesn't help much with changing the address or time period because `checkPriceHistory` is cheap. you also may get to a stale cache by not expiring it in this function

Reply:
so long story short - i don't think you need a useEffect here, but i do think you need some notion of `isFetching`